### ISSUE TITLE
The stdlib build currently only allows setting a single threading library type, even for multiple stdlibs

### ISSUE THREAD

**Issue by buttaface** [apple/swift#59568]
**Describe the bug**
Since #59287, it is possible to set a threading library for the Swift stdlib with `SWIFT_THREADING_PACKAGE`, which is [set to the host platform](https://github.com/apple/swift/pull/59287/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR197) by [default](https://github.com/apple/swift/pull/59287/files#diff-5c1eb2d6768a23753402274997290de28b4c44906856c32d858730083df6857eR196). However, [the current config builds all stdlibs with a single type of threading library](https://github.com/apple/swift/blob/f5bdb858e0e5a6476a00ff20db05cf74a0ded49c/cmake/modules/Threading.cmake#L23), which can be overridden with the flag `--swift-threading-package` but still applies to every stdlib. Since it's possible to cross-compile multiple non-host stdlibs, this won't work if any of them need a different threading library, though it's fine if they're all Darwin platforms with the same type of threading library, say macOS and iOS.

**To Reproduce**
1. Build multiple stdlibs that require different threading libraries, and watch it usually fail when all are built against the host threading library type.

**Expected behavior**
You should be able to build multiple stdlibs against different types of threading libraries.

**Environment:**
 - OS: Ubuntu 20.04 x86_64

**Additional context**
I just ran into this because the Android armv7 stdlib built fine on the community Android armv7 CI, but failed on my Android CI. I dug into it and it's because the host SDK on the community CI is `LINUX` so [it applies that threading library to Android armv7 too](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/lastFailedBuild/console), based on the default logic linked above. My own Android CI doesn't build for the linux host though, only for Android armv7 [with the flags `--cross-compile-hosts=android-armv7 --skip-local-build`](https://github.com/buttaface/swift-android-sdk/blob/main/.github/workflows/sdks.yml#L134), so the host default logic chooses pthreads for the Android host and [fails for armv7 only](https://github.com/buttaface/swift-android-sdk/runs/6933354497?check_suite_focus=true#step:6:1057), as [pthreads doesn't support 32-bit here](https://github.com/apple/swift/pull/59287/files#diff-01c32d9dd91f3dff021d98ee0357a5e7c9d0c059eaab416065de0dbe54d2040aR134).

In this case, it so happens that this issue is getting the Android build to pass, but in general, it's not a good idea to build all stdlibs against only one type of threading library.

**Comment by buttaface** [comment]
I don't think that pull is going to fix it, as you're still [only applying a single threading package to all stdlibs](https://github.com/apple/swift/pull/59589/files#diff-50b821298d876aad84496a736406c7adcac839dd49a321755906f69051ec9232R31). I have a couple questions:
1. You have some logic that checks [if the threading package variable is an empty string](https://github.com/apple/swift/pull/59589/files#diff-50b821298d876aad84496a736406c7adcac839dd49a321755906f69051ec9232R21), when would that be the case? [Since that variable is always default-initialized](https://github.com/apple/swift/pull/59589/files#diff-50b821298d876aad84496a736406c7adcac839dd49a321755906f69051ec9232R21), I don't think it will ever be empty, unless someone manually hardcodes it.
2. It appears you use that same variable for [the host](https://github.com/apple/swift/blob/main/cmake/modules/AddSwift.cmake#L312) and [the target stdlibs](https://github.com/apple/swift/blob/main/stdlib/cmake/modules/AddSwiftStdlib.cmake#L358), as both call `threading_package_name()`, which just uses [the host-initialized `SWIFT_THREADING_PACKAGE`](https://github.com/apple/swift/pull/59589/files#diff-50b821298d876aad84496a736406c7adcac839dd49a321755906f69051ec9232R31). How is that supposed to work for cross-compilation, are you expecting that `SWIFT_THREADING_PACKAGE` variable to be set to the empty string in that case?
3. Do you intend for the flag `--swift-threading-package` to also apply for the host and all targets? Because that is what it does now.

I'm unclear on some of this thread library selection logic, please let me know your intent.

**Comment by al45tair** [comment]
There is another variable I'd added, `SWIFT_SDK_${SDK}_THREADING_PACKAGE` that you can explicitly set, but that only gets used by the tests at present. Maybe that should also be used by the stdlib build, since it seems not entirely improbably that you might want to override it for some particular SDK.

**Comment by al45tair** [comment]
Right. I've updated my PR if you'd like to take a look (but please review it _there_, rather than here, otherwise things just get confusing).

**Comment by buttaface** [comment]
Confirmed that #59606 fixes this issue by trying it with various overrides, thanks for adding the list of overrides too.
